server/methods/core/cart.js

```diff
@@ -721,45 +743,95 @@ Meteor.methods({
       throw new Meteor.Error(404, "Cart not found",
         "Cart not found for user with such id");
     }
-
+    // temp hack until we build out multiple shipment handlers
+    // set the same address for every shipping record
     let selector;
     let update;
-    const primaryShopId = Reaction.getPrimaryShopId();
-    // temp hack until we build out multiple shipment handlers
-    // if we have an existing item update it, otherwise add to set.
-    if (Array.isArray(cart.shipping) && cart.shipping.length > 0) {
-      selector = {
-        "_id": cartId,
-        "shipping._id": cart.shipping[0]._id
-      };
-      update = {
-        $set: {
-          "shipping.$.address": address,
-          "shopId": primaryShopId
-        }
-      };
+    let updated = false;
+    if (cart.shipping && cart.shipping.length > 0 && cart.items) {
+      const shopIds = Object.keys(cart.getItemsByShop());
+      shopIds.map((shopId) => {
+        selector = {
+          "_id": cartId,
+          "shipping.shopId": shopId
+        };
+
+        update = {
+          $set: {
+            "shipping.$.address": address
+          }
+        };
+      });
+      try {
+        Collections.Cart.update(selector, update);
+        updated = true;
+      } catch (error) {
+        Logger.error("An error occurred adding the address", error);
+        throw new Meteor.Error("An error occurred adding the address", error);
+      }
     } else {
-      selector = {
-        _id: cartId
-      };
-      update = {
-        $addToSet: {
-          shipping: {
-            address: address,
-            shopId: primaryShopId
+      // if no items in cart just add or modify one record for the carts shop
+      if (!cart.items) {
+        // add a shipping record if it doesn't exist
+        if (!cart.shipping) {
+          selector = {
+            _id: cartId
+          };
+          update = {
+            $push: {
+              shipping: {
+                address: address,
+                shopId: cart.shopId
```

Comment:
If there is a default address stored in the addressbook it's added to the cart automatically on cart creation. If we don't add it here or update it then it will be lost when we add items to the cart. We could probably change this behavior since the way we use shipping records is sort of evolving.